### ISSUE TITLE
Figure out what to do with focusin/focusout

### ISSUE THREAD

**Issue by smaug----** [w3c/uievents#88]
What the spec defines is actually useful behavior, but unfortunately implementations do something totally different. (and that something else happens to be rather useless.) See for example https://github.com/jquery/jquery/issues/3123

@adrianba @RByers @jacobrossi @rniwa would blink/webkit/edge be interested in to fix their implementations, or should we change the spec? (I assume the latter)
There is also the option to remove the events altogether.

**Comment by RByers** [comment]
I don't know much about these events myself, but clearly fixing things to be interoperable one way or another is important.  @dtapuska / @choniong thoughts?

**Comment by dtapuska** [comment]
The ordering of the events focusout & blur,  focusin & focus could be made whatever we want as long as they are in pairs. The spec indicates an intermixed scenario which we don't implement.

The [relevant code](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/page/FocusController.cpp&sq=package:chromium&type=cs&l=309&rcl=1464247007) fires blur, focusout, DOMFocusOut and focus, focusin, DOMFocusIn.

We should probably also look at making DOMFocusOut/DOMFocusIn as legacy event mappings for focusin/focusout so that an event target doesn't actually get both.

**Comment by smaug----** [comment]
Let's not talk about DOMFocusOut/In here, given that they are by all means deprecated.
But it is the ordering of focusin, focusout, blur and focus which is really odd in some engines. IE seems to have them the right way.

I tried to look for webkit bugs implementing focusin/out, but couldn't find the relevant one.  That might have revealed why the behavior is so odd in webkit/blink.

The whole point of focusout/in is that they happen before blur/focus, but apparently that isn't happening, so the use cases for focusout/in are very weak.

**Comment by smaug----** [comment]
> The ordering of the events focusout & blur, focusin & focus could be made whatever we want as long as they are in pairs.

So that just doesn't make much sense, since, as I said
"The whole point of focusout/in is that they happen before blur/focus, but apparently that isn't happening, so the use cases for focusout/in are very weak."

@dtapuska I assume blink won't be changing its behavior? Same question to @rniwa and @jacobrossi
If so, we need to get whatever odd behavior is implemented spec'ed.

**Comment by smaug----** [comment]
@adrianba @rniwa  @dtapuska Do you think we should spec the rather useless ordering blink/webkit/edge now has, or would you be willing to fix implementations to follow the spec?
Or third option is to drop support for focusin/out.

**Comment by dtapuska** [comment]
I'm open to fixing blink. I'd like to hear from the edge team before blink would move on this. I think I'd like to see what jquery needs and provide that so they don't need to polyfill it.

**Comment by smaug----** [comment]
@mgol ^

**Comment by adrianba** [comment]
Adding @travisleithead.

**Comment by RByers** [comment]
If we change blink, with the existing jQuery code continue to work or will it be broken by the change?

**Comment by travisleithead** [comment]
I'm definitely in favor of interop, and I'd prefer to have the right (useful) event ordering here that makes the most sense. If we're going to change to a particular new ordering then it: should be done together (all implementations) and done with minimal web compatibility fallout.

> I think I'd like to see what jquery needs and provide that so they don't need to polyfill it.

Agree.

> If we change blink, with the existing jQuery code continue to work or will it be broken by the change?

Good question.

**Comment by smaug----** [comment]
@teddink

**Comment by dtapuska** [comment]
@travisleithead do you have any justification why the implementation changed between Edge and IE? I just want to understand if there were any interop concerns in this area.

**Comment by smaug----** [comment]
Looks like blink doesn't fire focusin/out on Window object, even though it does fire focus/blur.

**Comment by dmethvin** [comment]
We discussed this issue in the jQuery core team meeting today and our recommendation is:

1) Browsers should use the standards-mandated event order: focusout, focusin, blur, focus. Since focusout and focusin are cancellable they need to happen before focus/blur to prevent take-backsies. 

2) Browsers **must** provide some feature detect so that developers can know if focusin/out are implemented. Ideally this would be done at the same time as the first point so developers could know that the events are also implemented *per-spec*. At the moment we don't know of any reliable feature detect for these events.

The only thing jQuery guarantees right now is that focusin happens before focus and focusout happens before blur. Only IE11 has a reliable way to detect the events so jQuery simulates them even when the browser has them, which would be bad except that nobody seems to implement them correctly anyway. Once there is a reliable feature detect jQuery can defer to the working browser implementation.

**Comment by timmywil** [comment]
Another reason we like the spec order is related to something that's already been mentioned. We've had several tickets over the years about things like `.is(':focus')` not matching in a focus handler. It would be nice to have a handler for when the element is about to receive focus and one for when it has received focus. The spec handles that well.

**Comment by markelog** [comment]
Also see current state of impl in browsers at https://github.com/jquery/jquery/issues/3123#issue-155077612

**Comment by smaug----** [comment]
More weird stuff in blink. If focus is moved in blur handler, focusout it still dispatched, but if
focus is moved in focus handler, focusin isn't dispatched.
So, in blink it is possible to have events for the same target in order: blur, focus, focusin, focusout and yet have that target focused after focusout event.

**Comment by smaug----** [comment]
As of now I'm hoping to replicate most of blink's behavior in Gecko, since at least that behavior should be web compatible. But, since that model is quite odd, I'd be very happy if we all agreed to fix this and follow the model closer to what is in the current spec.  Unfortunately I haven't seen much related implementation work happening in other browsers, so expectations in that aren't too high.

**Comment by dtapuska** [comment]
@smaug----  I'd prefer that we actually shipped something that made sense. We have written interop tests here: https://github.com/w3c/web-platform-tests/pull/3857 (from @sunyunjia) 

We were waiting on hearing back from jquery. But it appears they were expecting cancelable events for focusin/focusout.

We are going to try to change our processing model this quarter but I'm not sure what monsters are lurking in the corners.

**Comment by dmethvin** [comment]
Also, focus-y events should work properly even when the browser window isn't in focus, that would be a good thing to ensure.

**Comment by markelog** [comment]
Perhaps you could provide a little bit of context - why it was decided to make `focus(in | out)` not cancelable?

**Comment by dmethvin** [comment]
It does seem like making them cancelable would be handy at times. Note that IE and Edge get focus management wrong when the focus leaves the window and actually prevent the user from even escaping to the browser chrome.

http://jsbin.com/kewuciyide/edit?html,output

**Comment by annevk** [comment]
@dtapuska any updates here?

**Comment by dtapuska** [comment]
We (Chrome) want to do this work although it isn't clear what interop issues we will have so it has gotten low priority. Is there new problems that this would solve which justifies a bump in priority?

**Comment by annevk** [comment]
@dtapuska basically, the longer we wait the less likely it becomes. It also would be good to have all focus events defined in detail to make sure that new platform features such as shadow roots interact well with them.

**Comment by patricia-gallardo** [comment]
Since my pull-request on uievents/order-of-events/focus-events/focus-manual.html in web-platform-tests is currently blocked because of this issue, I thought I'd chime in :) 
https://github.com/web-platform-tests/wpt/pull/9049#issuecomment-418642894

In my testing for that change and the accompanying change in the w3c html spec:
https://github.com/w3c/html/issues/278

it seems that most browsers have "naturally" converged on pretty much the same behavior which is reflected in my changes in the test. The current test only passes in IE and the behavior is not very intuitive imo. With my changes the test passes in Chrome, Firefox, Safari and almost in Edge. Since this seems to be the model the web has landed on I suggest changing the test rather than the implementers.

**Comment by mgol** [comment]
I re-checked browser behavior. Edge 17 & newer now matches Chrome. Firefox implemented that order as well. All dispatch the events in the following order:

1. `blur`
1. `focusout`
1. `focus`
1. `focusin`

IE 11 does what [the spec](https://www.w3.org/TR/uievents/#events-focusevent-event-order) mandates, i.e.:

1. `focusout`
1. `focusin`
1. `blur`
1. `focus`

The test case checking native behavior: https://jsfiddle.net/66znLhfw/15/

Is this still realistic to change the order when all current browsers implemented the same non-standard order?

**Comment by mfreed7** [comment]
I just hit this issue (again) and I'm curious where we stand. The [spec still says](https://w3c.github.io/uievents/#events-focusevent-event-order):

1. focusout
2. focusin
3. blur
4. focus

...but Blink, Gecko, and WebKit [**all** do this](https://jsfiddle.net/66znLhfw/15/):

1. blur
2. focusout
3. focus
4. focusin

While the specified behavior makes much more sense to me, I think that horse has left the barn, and we should likely update the spec to match the behavior. Thoughts?

**Comment by annevk** [comment]
I agree that we should specify what is implemented. And then if there is sufficient interest for a change a new proposal can be made for that and evaluated separately.

Having said that, I suggest that we take this opportunity to define these events in the HTML Standard, where focus is already mostly defined anyway: https://github.com/whatwg/html/issues/3514.

**Comment by smaug----** [comment]
I would be very surprised if this can be changed anymore, I mean from what has been shipping for years now.

**Comment by annevk** [comment]
#185 tracks moving all of this to the HTML Standard, for anyone wondering.
